spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala

```diff
@@ -206,36 +292,49 @@ case class DeltaSharingSource(
    * @return the last IndexedFile or None if there are no new data.
    */
   private def getLastFileChangeWithRateLimit(
-    fromVersion: Long,
-    fromIndex: Long,
-    isStartingVersion: Boolean,
-    limits: Option[AdmissionLimits] = Some(new AdmissionLimits())): Option[IndexedFile] = {
-    if (options.readChangeFeed) {
-      throw DeltaSharingErrors.CDFNotSupportedInStreaming
-    } else {
-      maybeGetFileChanges(fromVersion, fromIndex, isStartingVersion)
-
-      if (limits.isEmpty) return sortedFetchedFiles.lastOption
-
-      // Check each change until we've seen the configured number of addFiles. Return the last one
-      // for the caller to build offset.
-      var admissionControl = limits.get
-      var lastFileChange: Option[IndexedFile] = None
-      sortedFetchedFiles.foreach{indexedFile =>
-        if (admissionControl.admit(Option(indexedFile.add))) {
-          lastFileChange = Some(indexedFile)
-        } else {
-          return lastFileChange
+      fromVersion: Long,
+      fromIndex: Long,
+      isStartingVersion: Boolean,
+      limits: Option[AdmissionLimits] = Some(new AdmissionLimits())): Option[IndexedFile] = {
+    maybeGetFileChanges(fromVersion, fromIndex, isStartingVersion)
+
+    if (limits.isEmpty) return sortedFetchedFiles.lastOption
+
+    // Check each change until we've seen the configured number of addFiles. Return the last one
+    // for the caller to build offset.
+    var admissionControl = limits.get
+    var lastFileChange: Option[IndexedFile] = None
+    var index = 0
+    while (index < sortedFetchedFiles.size) {
+      var indexedFile = sortedFetchedFiles(index)
+
+      if (admissionControl.admit(indexedFile.getFileAction)) {
+        // For CDC commits we either admit the entire commit or nothing at all.
+        // This is to avoid returning `update_preimage` and `update_postimage` in separate
+        // batches.
+        while (indexedFile.cdc != null && index + 1 < sortedFetchedFiles.size
+          && sortedFetchedFiles(index + 1).cdc != null &&
+          sortedFetchedFiles(index + 1).version == indexedFile.version
+        ) {
+          // while is cdc file and on the same version, admit the file.
```

Comment:
Consider adding that we may exceed the admission control limits due to this, and consider explaining why that is acceptable.

Reply:
added in the comment above.